### PR TITLE
Reimplement BigInteger methods as functions

### LINKED ISSUE #17699: [Library Stabilization] BigInteger methods overwrite their receiver rather than returning
The documentation indicates that this is better for performance.  Should we provide standalone versions that return a new `bigint` instead/in addition? Note that a lot of methods like this create a local copy of their arguments when the arguments are remote (and usually there are two arguments).  Should we rename them?  Should we add compiler support for translating less optimal (but more understandable) forms into optimal forms?  I personally find it a bit odd, though I understand the reason behind it.

Michael says:
> I think it'd be reasonable to consider naming things like proc bigint.add differently [to make it clear that these overwrite their arguments]. But I don't know what we would name them to that would 1) be clear to newcomers and 2) be something one could easily map to a GMP operation.

Nikhil also brought this issue up at https://github.com/chapel-lang/chapel/issues/17616#issuecomment-834053841 and there was some good discussion in https://github.com/chapel-lang/chapel/issues/17616#issuecomment-834247288, https://github.com/chapel-lang/chapel/issues/17616#issuecomment-834544569, and https://github.com/chapel-lang/chapel/issues/17616#issuecomment-834622585

Affected methods are:
- `bigint.divexact()`
- `bigint.powm()`
- `bigint.pow()`
- `bigint.rootrem()`
- `bigint.sqrt()`
- `bigint.sqrtrem()`
- `bigint.nextprime()`
- `bigint.gcd()`
- `bigint.lcm()`
- `bigint.remove()`
- `bigint.gcdext()`
- `bigint.invert()`
- `bigint.fac()`
- `bigint.bin()`
- `bigint.fib()`
- `bigint.fib2()`
- `bigint.lucnum()`
- `bigint.lucnum2()`
- `bigint.add()`
- `bigint.sub()`
- `bigint.mul()`
- `bigint.addmul()`
- `bigint.submul()`
- `bigint.mul_2exp()`
- `bigint.neg()`
- `bigint.abs()`
- `bigint.div_q()`
- `bigint.div_r()`
- `bigint.div_qr()`
- `bigint.div_q_2exp()`
- `bigint.div_r_2exp()`
- `bigint.and()`
- `bigint.ior()`
- `bigint.xor()`
- `bigint.com()`

### PR DESCRIPTION
**Pull request by jabraham17**
Reimplements a number of methods that overwrite their receivers as standalone functions in the BigInteger module

## Summary of changes
- change methods to functions
  - `bigint.divexact()`
  - `bigint.powMod()`
  - `bigint.pow()`
  - `bigint.root()`
  - `bigint.rootrem()`
  - `bigint.sqrt()`
  - `bigint.sqrtrem()`
  - `bigint.nextprime()`
  - `bigint.gcd()`
  - `bigint.lcm()`
  - `bigint.invert()`
  - `bigint.removeFactor()`
  - `bigint.fac()`
  - `bigint.bin()`
  - `bigint.fib()`
  - `bigint.fib2()`
  - `bigint.lucnum()`
  - `bigint.lucnum2()`
  - `bigint.add()`
  - `bigint.sub()`
  - `bigint.mul()`
  - `bigint.addmul()`
  - `bigint.submul()`
  - `bigint.mul_2exp()`
  - `bigint.neg()`
  - `bigint.abs()`
  - `bigint.divQ()`
  - `bigint.divR()`
  - `bigint.divQR()`
  - `bigint.divQ2Exp()`
  - `bigint.divR2Exp()`
  - `bigint.mod()`
  - `bigint.and()`
  - `bigint.ior()`
  - `bigint.xor()`
  - `bigint.com()`

## New Tests
- deprecation tests for all methods that were deprecated

## testing
- full paratest
- full paratest + gasnet
- built docs locally

---
closes #17699
part of cray/chapel-private#4441

Note to reviewer: This PR contains a large number of changes. The majority are changes within BigInteger.chpl to rewrite methods as functions. The rest are updating the usage of these methods throughout the code base.

[Reviewed by @bmcdonald3 ]

### CODE REVIEW

File: modules/standard/BigInteger.chpl [hunk 1]
```diff
@@ -2767,88 +2825,95 @@ module BigInteger {
...
-      mpz_gcd(this.mpz, a.mpz, b.mpz);
-
-    } else if this.localeId == chpl_nodeID &&
+      mpz_gcd(result.mpz, a.mpz, b.mpz);
+    } else if result.localeId == chpl_nodeID &&
               a.localeId    == chpl_nodeID &&
               b.localeId    == chpl_nodeID {
-      mpz_gcd(this.mpz, a.mpz, b.mpz);
-
+      mpz_gcd(result.mpz, a.mpz, b.mpz);
```
Comment (bmcdonald3) [inline modules/standard/BigInteger.chpl:2854]: Looking at this three-pronged check and now that we have localize, I wonder if this could be something like

```chpl
if result.localeId == chpl_nodeID {
  a_ = a.localize();
  b_ = b.localize();
  mpz_gcd(result.mpz, a_.mpz, b_.mpz);
}
```

Which would result in less taking of the `else` path, which I think would be preferable here.

File: modules/standard/BigInteger.chpl [hunk 2]
```diff
@@ -2767,88 +2825,95 @@ module BigInteger {
...
-      mpz_nextprime(this.mpz, a.mpz);
-
-    } else if this.localeId == chpl_nodeID &&
-              a.localeId    == chpl_nodeID {
-      mpz_nextprime(this.mpz, a.mpz);
-
+      mpz_nextprime(result.mpz, a.mpz);
+    } else if result.localeId == chpl_nodeID &&
+              a.localeId == chpl_nodeID {
+      mpz_nextprime(result.mpz, a.mpz);
```
Comment (bmcdonald3) [inline modules/standard/BigInteger.chpl:2833]: Similar comment as below, let's change this to just check the result locale and then localize `a` so we take the `else` less.

File: modules/standard/BigInteger.chpl [hunk 3]
```diff
@@ -2767,88 +2825,95 @@ module BigInteger {
...
     if _local {
-      mpz_gcd_ui(this.mpz, a.mpz, b_);
-
-    } else if this.localeId == chpl_nodeID &&
+      mpz_gcd_ui(result.mpz, a.mpz, b_);
+    } else if result.localeId == chpl_nodeID &&
               a.localeId    == chpl_nodeID {
-      mpz_gcd_ui(this.mpz, a.mpz, b_);
-
+      mpz_gcd_ui(result.mpz, a.mpz, b_);
```
Comment (bmcdonald3) [inline modules/standard/BigInteger.chpl:2890]: Similar comment; if we could switch all of these multi-pronged checks to localize everything except for `result`, that would be good.

Could be worth splitting it into a separate PR so that this PR doesn't continue to grow and grow! (or you can just open an issue if you can't get to it now and we can get to it in time)

File: modules/standard/BigInteger.chpl [hunk 4]
```diff
@@ -4289,68 +4374,98 @@ module BigInteger {
...
     }
   }
 
-  // Note: Documentation on `denom: integral` version
-  proc bigint.divQR(ref       remain: bigint,
+  proc divQR(ref       result: bigint,
+                    ref       remain: bigint,
                     const ref numer: bigint,
                     const ref denom: bigint,
                     param     rounding = round.zero) {
```
Comment (bmcdonald3) [inline modules/standard/BigInteger.chpl:4389]: Looks like the spacing here isn't lining up anymore?

**Review by bmcdonald3 — APPROVED** [review]
Looks good, just one comment on spacing that I think needs to be addressed in this PR.

While here, I noticed 2 other bigger-picture things that I think should be addressed in separate PRs:
1. rethink the case where there are two bigint `ref`s to get the reslut 
    - I've opened https://github.com/Cray/chapel-private/issues/4628 for this
2. remove all the cases where we check more than just `result.localeId` and localize the other variables instead
    - this should definitely be a in a separate PR since it will also be a bigger change
    - let me know if you plan on getting to this, otherwise we can just capture it in an issue for now (either way is OK, no pressure)

**Comment by bmcdonald3** [comment]
Oops, and also the question about performance, but I'd be fine if we ignore that for now, since I don't expect the `localize()` call to have any impact (and possibly even a positive impact) upon further reflection.

**Review by bmcdonald3 — CHANGES_REQUESTED** [review]

**Comment by jabraham17** [comment]
Reverted all submitted benchmarks. Tested locally but rerunning paratest to make sure I didn't miss anything. This should be good to go now, and as soon as I merge this I will create a PR to take care of the extra checks beyond `result.localeId`.

**Review by bmcdonald3 — APPROVED** [review]

File: modules/standard/BigInteger.chpl [hunk 5]
```diff
@@ -3535,508 +3649,421 @@ module BigInteger {
...
-      mpz_mul_ui(this.mpz, a.mpz, b_);
+  @deprecated(notes="bigint.mul method is deprecated - please use the standalone function :proc:`~BigInteger.mul`")
+  proc bigint.mul(const ref a: bigint, b: int) {
+    BigInteger.mul(this, a, b);
+  }
 
-    } else if this.localeId == chpl_nodeID && a.localeId == chpl_nodeID {
-      mpz_mul_ui(this.mpz, a.mpz, b_);
+  proc mul(ref result: bigint, const ref a: bigint, b: uint) {
+    const b_ = b.safeCast(c_long);
```
Comment (bradcray) [inline modules/standard/BigInteger.chpl:3876]: @jabraham17 : Shouldn't this be doing a safe cast to `c_ulong` rather than `c_long`?

Comment (jabraham17) [reply modules/standard/BigInteger.chpl:3876]: Yes it should. I'll fix that right now

Comment (jabraham17) [reply modules/standard/BigInteger.chpl:3876]: This will be fixed by https://github.com/chapel-lang/chapel/pull/26295
